crates/sui-indexer-alt-jsonrpc/src/api/coin.rs

```diff
@@ -0,0 +1,349 @@
+
+use super::rpc_module::RpcModule;
+use crate::context::Context;
+use crate::data::objects::load_latest;
+use crate::error::{invalid_params, InternalContext, RpcError};
+use crate::paginate::{Cursor, Page};
+use diesel::dsl::sql;
+use diesel::prelude::*;
+use diesel::sql_types::Bool;
+use jsonrpsee::{core::RpcResult, proc_macros::rpc};
+use move_core_types::language_storage::TypeTag;
+use serde::{Deserialize, Serialize};
+use sui_indexer_alt_schema::objects::StoredCoinOwnerKind;
+use sui_indexer_alt_schema::schema::coin_balance_buckets;
+use sui_json_rpc_types::{Coin, Page as JsonRpcPage};
+use sui_open_rpc::Module;
+use sui_open_rpc_macros::open_rpc;
+use sui_types::{
+    base_types::{ObjectID, SuiAddress},
+    gas_coin::GAS,
+    object::Object,
+};
+
+#[open_rpc(namespace = "suix", tag = "Coin API")]
+#[rpc(server, namespace = "suix")]
+trait CoinsApi {
+    /// Return Coin objects owned by an address with a specified coin type.
+    /// If no coin type is specified, SUI coins are returned.
+    #[method(name = "getCoins")]
+    async fn get_coins(
+        &self,
+        /// the owner's Sui address
+        owner: SuiAddress,
+        /// optional coin type
+        coin_type: Option<String>,
+        /// optional paging cursor
+        cursor: Option<String>,
+        /// maximum number of items per page
+        limit: Option<usize>,
+    ) -> RpcResult<JsonRpcPage<Coin, String>>;
+
+    /// Return all Coin objects owned by an address.
+    #[method(name = "getAllCoins")]
+    async fn get_all_coins(
+        &self,
+        /// the owner's Sui address
+        owner: SuiAddress,
+        /// optional paging cursor
+        cursor: Option<String>,
+        /// maximum number of items per page
+        limit: Option<usize>,
+    ) -> RpcResult<JsonRpcPage<Coin, String>>;
+}
+
+pub(crate) struct Coins(pub Context, pub CoinsConfig);
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CoinsConfig {
+    /// The default page size limit when querying coins, if none is provided.
+    pub default_page_size: usize,
+
+    /// The largest acceptable page size when querying coins. Requesting a page larger than
+    /// this is a user error.
+    pub max_page_size: usize,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub(crate) enum Error {
+    #[error("Pagination issue: {0}")]
+    Pagination(#[from] crate::paginate::Error),
+
+    #[error("Failed to parse type {0:?}: {1}")]
+    BadType(String, anyhow::Error),
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct BalanceCursor {
+    #[serde(rename = "o")]
+    object_id: ObjectID,
+    #[serde(rename = "c")]
+    cp_sequence_number: i64,
+    #[serde(rename = "b")]
+    coin_balance_bucket: i16,
+}
+
+#[async_trait::async_trait]
+impl CoinsApiServer for Coins {
+    async fn get_coins(
+        &self,
+        owner: SuiAddress,
+        coin_type: Option<String>,
+        cursor: Option<String>,
+        limit: Option<usize>,
+    ) -> RpcResult<JsonRpcPage<Coin, String>> {
```

Comment:
nit: for consistency with other implementations of paginated methods (and also generally with the naming scheme for response types generally).

```suggestion
    ) -> RpcResult<PageResponse<Coin, String>> {
```